packages/editor/src/components/block-removal-warnings/index.js

```diff
@@ -0,0 +1,85 @@
+/**
+ * WordPress dependencies
+ */
+
+import { _n } from '@wordpress/i18n';
+import { privateApis as blockEditorPrivateApis } from '@wordpress/block-editor';
+import { useSelect } from '@wordpress/data';
+import { useMemo } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { unlock } from '../../lock-unlock';
+import { store as editorStore } from '../../store';
+
+const { BlockRemovalWarningModal } = unlock( blockEditorPrivateApis );
+
+// Prevent accidental removal of certain blocks, asking the user for confirmation first.
+const TEMPLATE_BLOCKS = [
+	'core/post-content',
+	'core/post-template',
+	'core/query',
+];
+const BLOCK_REMOVAL_RULES = [
+	{
+		// Template blocks.
+		// The warning is only shown when a user manipulates templates or template parts.
+		postTypes: [ 'wp_template', 'wp_template_part' ],
+		callback( removedBlocks ) {
+			const removedTemplateBlocks = removedBlocks.filter( ( { name } ) =>
+				TEMPLATE_BLOCKS.includes( name )
+			);
+			if ( removedTemplateBlocks.length ) {
+				return _n(
+					'Deleting this block will stop your post or page content from displaying on this template. It is not recommended.',
+					'Some of the deleted blocks will stop your post or page content from displaying on this template. It is not recommended.',
```

Comment:
I revised the copy that was in `trunk` as we don't know that all of the deleted blocks are dangerous to be removed, it could be four paragraphs and one query loop.

We could also go through all the permutations to make sure the messaging is completely correct, but I thought this would be sufficient.